src/arithmetic/func_desc.c

```diff
@@ -62,3 +66,32 @@ bool AR_FuncExists(const char *func_name) {
 	return (AR_GetFunc(func_name) != NULL);
 }
 
+inline void AR_SetPrivateDataRoutines(AR_FuncDesc *func_desc, AR_Func_Free bfree,
+									  AR_Func_Clone bclone) {
+	func_desc->bfree = bfree;
+	func_desc->bclone = bclone;
+}
+
+void AR_SetPrivateData(AR_FuncDesc **func_ptr, void *privdata) {
+	// Clone the function descriptor.
+	AR_FuncDesc *func = rm_malloc(sizeof(AR_FuncDesc));
+	memcpy(func, *func_ptr, sizeof(AR_FuncDesc));
+
+	// Set the private data pointer.
+	func->privdata = privdata;
+
+	// Replace the given function descriptor with the clone.
+	*func_ptr = func;
+}
+
+AR_FuncDesc *AR_CloneFuncDesc(const AR_FuncDesc *orig) {
+	ASSERT(orig->privdata && orig->bclone);
```

Comment:
That's fair, but I will keep the assertion on the function pointer as otherwise we'd need a conditional guard on the `orig->bclone(orig->privdata)` call.